// react-kronos/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _jss = require('jss');

var _jss2 = _interopRequireDefault(_jss);

var _jssNested = require('jss-nested');

var _jssNested2 = _interopRequireDefault(_jssNested);

var _jssCamelCase = require('jss-camel-case');

var _jssCamelCase2 = _interopRequireDefault(_jssCamelCase);

var _jssVendorPrefixer = require('jss-vendor-prefixer');

var _jssVendorPrefixer2 = _interopRequireDefault(_jssVendorPrefixer);

var _jssDefaultUnit = require('jss-default-unit');

var _jssDefaultUnit2 = _interopRequireDefault(_jssDefaultUnit);

var _constants = require('./constants');

var _calendar = require('./calendar');

var _calendar2 = _interopRequireDefault(_calendar);

var _styledComponent = require('./styled-component');

var _styledComponent2 = _interopRequireDefault(_styledComponent);

var _styles = require('./styles');

var _styles2 = _interopRequireDefault(_styles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

_jss2.default.use((0, _jssNested2.default)(), (0, _jssCamelCase2.default)(), (0, _jssVendorPrefixer2.default)(), (0, _jssDefaultUnit2.default)());

var ISOregex = /((\d{4}\-\d\d\-\d\d)[tT]([\d:\.]*)?)([zZ]|([+\-])(\d\d):?(\d\d))/;
var minutesOfDay = function minutesOfDay(m) {
  return (0, _moment2.default)(m).minutes() + (0, _moment2.default)(m).hours() * 60;
};

var Kronos = function (_Component) {
  _inherits(Kronos, _Component);

  function Kronos(props) {
    _classCallCheck(this, Kronos);

    var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(Kronos).call(this, props));

    _this.state = {
      datetime: _this.getDateTimeInput().datetime,
      input: _this.getDateTimeInput().input,
      type: _this.getDateTimeInput().type,
      visible: false,
      level: _this.getDefaultLevel()
    };
    return _this;
  }

  _createClass(Kronos, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (this.props != nextProps) {
        this.validate(this.getDateTimeInput(nextProps).datetime, null, true);
        this.setState({
          datetime: this.getDateTimeInput(nextProps).datetime,
          input: this.getDateTimeInput(nextProps).input
        });
      }
    }
  }, {
    key: 'getDateTimeInput',
    value: function getDateTimeInput(props) {
      props = props || this.props;
      var prop = props.date || props.time || null;

      var datetime = void 0,
          input = void 0,
          type = void 0;
      if (prop === null) {
        datetime = (0, _moment2.default)();
        input = null;
        type = _constants.Types.MOMENT;
      } else {
        datetime = this.parse(prop);
        input = datetime.format(this.format(props));
        switch (typeof prop === 'undefined' ? 'undefined' : _typeof(prop)) {
          case 'object':
            if (_moment2.default.isDate(prop)) {
              type = _constants.Types.JS_DATE;
            } else if (_moment2.default.isMoment(prop)) {
              type = _constants.Types.MOMENT;
            } else {
              type = null;
            }
            break;
          case 'string':
            if (prop.match(ISOregex)) {
              type = _constants.Types.ISO;
            } else {
              type = _constants.Types.STRING;
            }
            break;
        }
      }

      return {
        datetime: datetime,
        input: input,
        type: type
      };
    }
  }, {
    key: 'getDefaultLevel',
    value: function getDefaultLevel() {
      if (typeof this.props.date !== 'undefined') {
        return _constants.Units.DAY;
      } else if (typeof this.props.time !== 'undefined') {
        return _constants.Units.HOUR;
      } else {
        console.warn('Please set a date or time prop. It can be null but not undefined.');
        return _constants.Units.DAY;
      }
    }
  }, {
    key: 'format',
    value: function format(props) {
      props = props || this.props;
      if (typeof props.format !== 'undefined') {
        return props.format;
      } else if (typeof props.date !== 'undefined') {
        return 'MM-DD-YYYY';
      } else if (typeof props.time !== 'undefined') {
        return 'h:mm a';
      } else {
        return null;
      }
    }
  }, {
    key: 'toggle',
    value: function toggle(visible) {
      if (typeof visible === 'undefined') {
        visible = !this.state.visible;
      }
      if (visible !== this.state.visible) {
        this.setState({ visible: visible });
      }
    }
  }, {
    key: 'parse',
    value: function parse(input) {
      if (input === null) return null;
      var parsing = (0, _moment2.default)(input, this.format(), true);
      if (!parsing.isValid()) {
        var test = new Date(input);
        if (isNaN(test.getTime())) {
          test = this.state && this.state.datetime || (0, _moment2.default)();
        }

        parsing = (0, _moment2.default)(test);
      }

      return parsing;
    }
  }, {
    key: 'save',
    value: function save(saving) {
      var datetime = this.state.datetime;

      if (typeof this.props.date !== 'undefined') {
        saving.hours(datetime.hours());
        saving.minutes(datetime.minutes());
      }
      if (typeof this.props.time !== 'undefined') {
        saving.date(datetime.date());
        saving.month(datetime.month());
        saving.year(datetime.year());
      }
      this.setState({
        datetime: saving,
        input: saving.format(this.format())
      });

      if (this.validate(saving, null, true)) this.commit(saving);
    }
  }, {
    key: 'validate',
    value: function validate(datetime, timeUnit, isSaving) {
      var outsideRange = false;

      if (this.props.min && (0, _moment2.default)(datetime).isBefore(this.props.min)) {
        outsideRange = true;
      }
      if (this.props.max && (0, _moment2.default)(datetime).isAfter(this.props.max)) {
        outsideRange = true;
      }

      if (this.props.minTime && minutesOfDay(datetime) < minutesOfDay(this.props.minTime)) {
        outsideRange = true;
      }
      if (this.props.maxTime && minutesOfDay(datetime) > minutesOfDay(this.props.maxTime)) {
        outsideRange = true;
      }

      if (outsideRange && timeUnit !== 'hours') {
        if ((0, _moment2.default)(datetime).isSame(this.props.min, timeUnit) || (0, _moment2.default)(datetime).isSame(this.props.max, timeUnit)) {
          outsideRange = false;
        }
      }

      if (isSaving) {
        this.setState({ dateTimeExceedsValidRange: outsideRange });
        if (this.props.shouldTriggerOnChangeForDateTimeOutsideRange) return true;
      }

      return !outsideRange;
    }
  }, {
    key: 'commit',
    value: function commit(datetime) {
      var returnAs = this.props.returnAs || this.state.type;
      var result = void 0;
      switch (returnAs) {
        case _constants.Types.ISO:
          result = datetime.toISOString();
          break;
        case _constants.Types.JS_DATE:
          result = datetime.toDate();
          break;
        case _constants.Types.MOMENT:
          result = datetime;
          break;
        case _constants.Types.STRING:
          result = datetime.format(this.format());
          break;
      }

      this.props.onChangeDateTime && this.props.onChangeDateTime(result);
    }
  }, {
    key: 'onClickInput',
    value: function onClickInput(e) {
      if (this.props.onClick) this.props.onClick(e);
      this.toggle(true);
    }
  }, {
    key: 'onFocusInput',
    value: function onFocusInput(e) {
      if (this.props.onFocus) this.props.onFocus(e);
      this.toggle(true);
    }
  }, {
    key: 'onBlurInput',
    value: function onBlurInput(e) {
      var datetime = this.state.datetime || (0, _moment2.default)();

      if (this.above) {
        _reactDom2.default.findDOMNode(this._input).focus();
      } else if (this.props.closeOnBlur) {
        this.toggle(false);
        if (this.props.onBlur) this.props.onBlur(e);
      }
      if (this.state.input == this.state.datetime.format(this.format())) {
        return;
      } else {
        datetime = this.parse(this.state.input);
        if (datetime) this.save(datetime);
      }
    }
  }, {
    key: 'onChangeInput',
    value: function onChangeInput(e) {
      if (this.props.onChange) this.props.onChange(e);

      var input = e.target.value;
      var datetime = (0, _moment2.default)(input, this.format(), true);
      if (datetime.isValid()) {
        this.save(datetime);
      } else if (input == '') {
        this.setState({
          datetime: null,
          input: ''
        });
        this.props.onChangeDateTime && this.props.onChangeDateTime(null);
      } else {
        this.setState({ input: input });
      }
    }
  }, {
    key: 'onSelect',
    value: function onSelect(datetime, close, timeUnit) {
      var shouldClose = close;
      var visible = this.state.visible;
      var _props = this.props;
      var closeOnSelect = _props.closeOnSelect;
      var preventClickOnDateTimeOutsideRange = _props.preventClickOnDateTimeOutsideRange;


      if (timeUnit) {
        if (!this.validate(datetime, timeUnit.unit)) shouldClose = false;
      } else {
        if (!this.validate(datetime)) shouldClose = false;
      }
      if (close && shouldClose === false && preventClickOnDateTimeOutsideRange) return;

      var willBeVisible = closeOnSelect && shouldClose ? !visible : visible;

      this.setState({ visible: willBeVisible });

      this.save(datetime);

      if (this.props.onSelect) this.props.onSelect(datetime, willBeVisible);
    }
  }, {
    key: 'onKeyDown',
    value: function onKeyDown(code) {
      var datetime = this.state.datetime || (0, _moment2.default)();
      var lvl = _constants.Levels[this.state.level];

      switch (code) {
        case _constants.Keys.UP:
          this.onSelect(datetime.subtract(lvl.key.span, lvl.key.unit));
          break;
        case _constants.Keys.DOWN:
          this.onSelect(datetime.add(lvl.key.span, lvl.key.unit));
          break;
        case _constants.Keys.ENTER:
          if (lvl.down) {
            this.setState({ level: lvl.down });
          } else {
            if (this.state.input == datetime.format(this.format())) {
              if (!this.validate(datetime)) {
                this.toggle(true);
              } else {
                this.toggle();
              }
            } else {
              if (!this.state.visible) this.toggle(true);
              datetime = this.parse(this.state.input);
              if (datetime) this.save(datetime);
            }
          }
          break;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var mainClasses = (0, _classnames2.default)('react-kronos', this.props.id, this.props.classes.kronos);
      var inputClasses = (0, _classnames2.default)(this.props.classes.input, { 'outside-range': this.state.dateTimeExceedsValidRange });
      var visible = this.props.controlVisibility ? this.props.visible : this.state.visible;

      return _react2.default.createElement(
        'div',
        { className: mainClasses },
        _react2.default.createElement('input', {
          type: 'text',
          ref: function ref(input) {
            return _this2._input = input;
          },
          value: this.state.input || '',
          onClick: this.onClickInput.bind(this),
          onFocus: this.onFocusInput.bind(this),
          onBlur: this.onBlurInput.bind(this),
          onKeyDown: function onKeyDown(e) {
            return _this2.onKeyDown(e.keyCode);
          },
          onChange: this.onChangeInput.bind(this),
          placeholder: this.props.placeholder,
          name: this.props.name,
          className: inputClasses
        }),
        visible && _react2.default.createElement(_calendar2.default, {
          id: this.props.id,
          datetime: this.state.datetime,
          onSelect: this.onSelect.bind(this),
          above: function above(bool) {
            return typeof bool === 'undefined' ? _this2.above : _this2.above = bool;
          },
          level: this.state.level,
          setLevel: function setLevel(level) {
            return _this2.setState({ level: level });
          },
          validate: this.validate.bind(this),
          options: this.props.options,
          input: this._input
        })
      );
    }
  }]);

  return Kronos;
}(_react.Component);

Kronos.propTypes = {
  date: _react.PropTypes.any,
  time: _react.PropTypes.any,
  min: _react.PropTypes.any,
  max: _react.PropTypes.any,
  minTime: _react.PropTypes.any,
  maxTime: _react.PropTypes.any,
  shouldTriggerOnChangeForDateTimeOutsideRange: _react.PropTypes.bool,
  preventClickOnDateTimeOutsideRange: _react.PropTypes.bool,
  format: _react.PropTypes.string,
  onChangeDateTime: _react.PropTypes.func,
  returnAs: _react.PropTypes.oneOf([_constants.Types.ISO, _constants.Types.JS_DATE, _constants.Types.MOMENT, _constants.Types.STRING]),
  closeOnSelect: _react.PropTypes.bool,
  closeOnBlur: _react.PropTypes.bool,
  placeholder: _react.PropTypes.string,
  name: _react.PropTypes.string,
  options: _react.PropTypes.object,
  // Advanced controls
  controlVisibility: _react.PropTypes.bool,
  visible: _react.PropTypes.bool,
  onClick: _react.PropTypes.func,
  onFocus: _react.PropTypes.func,
  onBlur: _react.PropTypes.func,
  onChange: _react.PropTypes.func,
  onSelect: _react.PropTypes.func
};
Kronos.defaultProps = {
  closeOnSelect: true,
  closeOnBlur: true,
  controlVisibility: false,
  shouldTriggerOnChangeForDateTimeOutsideRange: false,
  preventClickOnDateTimeOutsideRange: false,
  visible: false
};
Kronos.above = false;
exports.default = (0, _styledComponent2.default)(Kronos, function (props, id) {
  return (0, _styles2.default)('index', props, id);
});
